src/libraries/System.IO.FileSystem/tests/File/ReadWriteAllLinesAsync.cs

```diff
@@ -199,4 +199,103 @@ public override Task TaskAlreadyCanceledAsync()
+
+        [Fact]
+        public override async Task TaskAlreadyCanceledAsync()
+        {
+            string path = GetTestFilePath();
+            File.Create(path).Dispose();
+            var ct = new CancellationToken(true);
+            await Assert.ThrowsAsync<TaskCanceledException>(() => ReadAsync(path, ct, default));
+            await Assert.ThrowsAsync<TaskCanceledException>(() => ReadAsync(path, default, ct));
+        }
+
+        [Fact]
+        public virtual void InvalidArgumentsThrowsBeforeGetAsyncEnumerator()
+        {
+            Assert.Throws<ArgumentNullException>("path", () => File.ReadLinesAsync(null));
+            Assert.Throws<ArgumentException>("path", () => File.ReadLinesAsync(string.Empty));
+        }
+
+        protected virtual async Task<string[]> ReadAsync(string path, CancellationToken enumerableCt, CancellationToken enumeratorCt)
+        {
+            var list = new List<string>();
+            await foreach (string item in File.ReadLinesAsync(path, enumerableCt).WithCancellation(enumeratorCt))
+            {
+                list.Add(item);
+            }
+
+            return list.ToArray();
+        }
+    }
+
+    public class File_ReadLines_EncodedAsync : File_ReadLinesAsync
+    {
+        [Fact]
+        public override void InvalidArgumentsThrowsBeforeGetAsyncEnumerator()
+        {
+            string path = GetTestFilePath();
+            Assert.Throws<ArgumentNullException>(() => File.ReadLinesAsync(path, null));
+            Assert.Throws<ArgumentNullException>("path", () => File.ReadLinesAsync(null, new UTF8Encoding(false)));
+            Assert.Throws<ArgumentException>("path", () => File.ReadLinesAsync(string.Empty, new UTF8Encoding(false)));
+        }
+
+        [Fact]
+        public override async Task DisposingEnumeratorClosesFileAsync()
+        {
+            string path = GetTestFilePath();
+            await WriteAsync(path, new[] { "line1", "line2", "line3" });
+
+            IAsyncEnumerable<string> readLines = File.ReadLinesAsync(path, new UTF8Encoding(false));
+            await using (IAsyncEnumerator<string> e1 = readLines.GetAsyncEnumerator())
+            await using (IAsyncEnumerator<string> e2 = readLines.GetAsyncEnumerator())
+            {
+                Assert.Same(readLines, e1);
+                Assert.NotSame(e1, e2);
+
+                await e1.MoveNextAsync();
+            }
+
+            // File should be closed deterministically; this shouldn't throw.
+            await File.OpenWrite(path).DisposeAsync();
+        }
+
+        protected override async Task<string[]> ReadAsync(string path, CancellationToken enumerableCt, CancellationToken enumeratorCt)
+        {
+            var list = new List<string>();
+            await foreach (string item in File.ReadLinesAsync(path, new UTF8Encoding(false), enumerableCt).WithCancellation(enumeratorCt))
+            {
+                list.Add(item);
+            }
+
+            return list.ToArray();
+        }
+    }
```

Comment:
We should be able to reduce code duplication:

```suggestion
    public class File_ReadLinesAsync : File_ReadWriteAllLines_EnumerableAsync
    {
        protected override Task<string[]> ReadAsync(string path)
            => ReadAsync(path, default, default);
        
        protected virtual IAsyncEnumerable<string> ReadLinesAsync(string path, CancellationToken ct = default)
            => File.ReadLinesAsync(path, ct);
            
        private async Task<string[]> ReadLinesAsync(string path, CancellationToken enumerableCt, CancellationToken enumeratorCt)
        {
            var list = new List<string>();
            await foreach (string item in ReadLinesAsync(path, enumerableCt).WithCancellation(enumeratorCt))
            {
                list.Add(item);
            }

            return list.ToArray();
        }

        [Fact]
        public override async Task DisposingEnumeratorClosesFileAsync()
        {
            string path = GetTestFilePath();
            await WriteAsync(path, new[] { "line1", "line2", "line3" });

            IAsyncEnumerable<string> readLines = ReadLinesAsync(path);
            await using (IAsyncEnumerator<string> e1 = readLines.GetAsyncEnumerator())
            await using (IAsyncEnumerator<string> e2 = readLines.GetAsyncEnumerator())
            {
                Assert.Same(readLines, e1);
                Assert.NotSame(e1, e2);

                await e1.MoveNextAsync();
            }

            // File should be closed deterministically; this shouldn't throw.
            await File.OpenWrite(path).DisposeAsync();
        }

        [Fact]
        public override async Task TaskAlreadyCanceledAsync()
        {
            string path = GetTestFilePath();
            File.Create(path).Dispose();
            var ct = new CancellationToken(true);
            await Assert.ThrowsAsync<TaskCanceledException>(() => ReadLinesAsync(path, ct));
            await Assert.ThrowsAsync<TaskCanceledException>(() => ReadAsync(path, ct, default));
            await Assert.ThrowsAsync<TaskCanceledException>(() => ReadAsync(path, default, ct));
        }

        [Fact]
        public void InvalidArgumentsThrowsBeforeGetAsyncEnumerator()
        {
            Assert.Throws<ArgumentNullException>("path", () => ReadLinesAsync(null));
            Assert.Throws<ArgumentException>("path", () => ReadLinesAsync(string.Empty));
        }
    }

    public class File_ReadLines_EncodedAsync : File_ReadLinesAsync
    {
        protected override IAsyncEnumerable<string> ReadLinesAsync(string path, CancellationToken ct = default)
            => File.ReadLinesAsync(path, new UTF8Encoding(false), ct);

        [Fact]
        public void InvalidArgumentsThrownForNullEncoding()
            => Assert.Throws<ArgumentNullException>("encoding", () => File.ReadLinesAsync("path", null));
    }
```

please take it with a grain of salt as I've not compiled it

Reply:
Got it, fixed, thanks.
Just note I am not adding suggested line 248:` await Assert.ThrowsAsync<TaskCanceledException>(() => ReadLinesAsync(path, ct)); ` because `ReadLinesAsync(path, ct)` is a sync call so current `ReadLinesAsync()` not check `cancellationToken` status.